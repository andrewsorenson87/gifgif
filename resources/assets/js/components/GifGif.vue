<template>
    <div>
        <div class="rounded shadow text-grey-darker">
            <div class="py-8">
                <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>

                <div class="flex justify-center py-8">
                    <label for="gif1" class="mx-8">
                        GIF
                        <input type="radio" id="gif1" value="gif1" name="pronunciation" v-model="newVote.picked" />
                    </label>

                    <label for="gif2" class="mx-8">
                        GIF
                        <input type="radio" id="gif2" value="gif2" name="pronunciation" v-model="newVote.picked" />
                    </label>
                </div>
            </div>

            <div class="flex justify-center">
                <div class="w-3/4">
                    <input class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight"
                           id="reason"
                           type="text"
                           v-model="newVote.reason"
                           placeholder="Why do you think it should be pronounced that way?" />
                </div>
            </div>

            <div class="flex justify-center my-4">
                <div class="w-3/4">
                    <input class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight"
                           id="name"
                           type="text"
                           v-model="newVote.name"
                           placeholder="What is your name?" />
                </div>
            </div>

            <div class="flex justify-center">
                <button id="submit"
                        v-bind:class="valid ? '' : 'opacity-50 cursor-not-allowed'"
                        :disabled="! valid"
                        class="bg-blue hover:bg-blue-dark text-white font-bold py-2 px-4 rounded my-8"
                        @click="add">
                    Submit
                </button>
            </div>
        </div>

        <div class="py-8">
            <h4 class="text-grey-darker text-center">User Responses:</h4>
            <ul v-for="vote in votes">
                <li class="py-4">
                    <p><span class="font-light uppercase">gif</span> â€”
                        {{ vote.reason }}</p>
                    <p class="text-sm mt-2">Submitted by <span class="uppercase">{{ vote.name }}</span> on 2018-07-25</p>
                </li>
            </ul>
        </div>
    </div>
</template>

<script>
    import axios from 'axios';

    export default {
        data() {
            return {
                newVote: {
                    picked: '',
                    reason: '',
                    name: ''
                },
                votes: []
            }
        },

        computed: {
            valid() {
                return this.newVote.picked && this.newVote.reason && this.newVote.name;
            }
        },

        methods: {
            add() {
                this.votes.push(this.newVote);

                this.newVote = {
                    picked: '',
                    reason: '',
                    name: ''
                };
            }
        }
    }
</script>
