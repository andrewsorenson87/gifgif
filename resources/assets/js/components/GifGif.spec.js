import GifGif from '@/GifGif.vue';
import { shallowMount } from '@vue/test-utils';

window.axios = {
    post: jest.fn().mockImplementation(() => Promise.resolve([])),
};

describe('GifGif', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(GifGif, {
            propsData: {
                dataVotes: [],
            }
        });
    });

    it('renders', () => {
        expect(wrapper.text()).toContain('How do you pronounce the word gif?');
        expect(wrapper).toMatchSnapshot();
    });

    it('shows votes loaded from props', () => {
        let reason = 'It is not jrafical interchange format';
        let name = 'Matt Stauffer';

        wrapper = shallowMount(GifGif, {
            propsData: {
                dataVotes: [{
                    picked: 'gif2',
                    reason,
                    name,
                }]
            }
        });

        expect(wrapper.text()).toContain(reason);
        expect(wrapper.text()).toContain(name);
        expect(wrapper).toMatchSnapshot();
    });

    it('updates the selected pronunciation via radio button', () => {
        wrapper.find('#gif1').trigger('click');
        expect(wrapper.vm.newVote.picked).toBe('gif1');

        wrapper.find('#gif2').trigger('click');
        expect(wrapper.vm.newVote.picked).not.toBe('gif1');
        expect(wrapper.vm.newVote.picked).toBe('gif2');
    });

    it('submits a reason for the pronunciation', async () => {
        let vote = {
            picked: 'gif2',
            reason: 'All the cool kids say it that way',
            name: 'Samantha',
        };

        expect(wrapper.vm.votes.length).toBe(0);

        wrapper.find('#gif2').trigger('click');
        expect(wrapper.vm.newVote.picked).toBe(vote.picked);

        let reason = wrapper.find('#reason');
        reason.element.value = vote.reason;
        reason.trigger('input');
        expect(wrapper.vm.newVote.reason).toBe(vote.reason);

        let name = wrapper.find('#name');
        name.element.value = vote.name;
        name.trigger('input');
        expect(wrapper.vm.newVote.name).toBe(vote.name);

        wrapper.find('#submit').trigger('click');

        expect(axios.post).toBeCalledWith('/api/votes', vote);

        await wrapper.vm.$nextTick();
        expect(wrapper.vm.votes.length).toBe(1);
        expect(wrapper.vm.newVote.picked).not.toBe('gif1');
        expect(wrapper.vm.newVote.reason).not.toBe(vote.reason);
        expect(wrapper.vm.newVote.name).not.toBe(vote.name);
        expect(wrapper).toMatchSnapshot();
    });

    it('does not let the user submit the form without all the fields', () => {
        wrapper.find('#submit').trigger('click');

        expect(wrapper.vm.votes.length).toBe(0);
    });
});
