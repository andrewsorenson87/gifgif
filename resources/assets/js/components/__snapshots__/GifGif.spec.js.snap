// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGif renders 1`] = `
<div>
  <div class="rounded shadow text-grey-darker">
    <div class="py-8">
      <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>
      <div class="flex justify-center py-8">
        <label for="gif1" class="mx-8">
          GIF
          <input type="radio" id="gif1" value="gif1" name="pronunciation">
        </label>
        <label for="gif2" class="mx-8">
          GIF
          <input type="radio" id="gif2" value="gif2" name="pronunciation">
        </label>
      </div>
    </div>
    <div class="flex justify-center">
      <div class="w-3/4">
        <input id="reason" type="text" placeholder="Why do you think it should be pronounced that way?" class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight">
      </div>
    </div>
    <div class="flex justify-center my-4">
      <div class="w-3/4">
        <input id="name" type="text" placeholder="What is your name?" class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight">
      </div>
    </div>
    <div class="flex justify-center">
      <button id="submit" disabled="disabled" class="bg-blue hover:bg-blue-dark text-white font-bold py-2 px-4 rounded my-8 opacity-50 cursor-not-allowed">
        Submit
      </button>
    </div>
  </div>
  <div class="py-8">
    <h4 class="text-grey-darker text-center">User Responses:</h4>
  </div>
</div>
`;

exports[`GifGif shows votes loaded from props 1`] = `
<div>
  <div class="rounded shadow text-grey-darker">
    <div class="py-8">
      <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>
      <div class="flex justify-center py-8">
        <label for="gif1" class="mx-8">
          GIF
          <input type="radio" id="gif1" value="gif1" name="pronunciation">
        </label>
        <label for="gif2" class="mx-8">
          GIF
          <input type="radio" id="gif2" value="gif2" name="pronunciation">
        </label>
      </div>
    </div>
    <div class="flex justify-center">
      <div class="w-3/4">
        <input id="reason" type="text" placeholder="Why do you think it should be pronounced that way?" class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight">
      </div>
    </div>
    <div class="flex justify-center my-4">
      <div class="w-3/4">
        <input id="name" type="text" placeholder="What is your name?" class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight">
      </div>
    </div>
    <div class="flex justify-center">
      <button id="submit" disabled="disabled" class="bg-blue hover:bg-blue-dark text-white font-bold py-2 px-4 rounded my-8 opacity-50 cursor-not-allowed">
        Submit
      </button>
    </div>
  </div>
  <div class="py-8">
    <h4 class="text-grey-darker text-center">User Responses:</h4>
    <ul>
      <li class="py-4">
        <p><span class="font-light uppercase">gif</span> — It is not jrafical interchange format</p>
        <p class="text-sm mt-2">Submitted by <span class="uppercase">Matt Stauffer</span> on 2018-07-25</p>
      </li>
    </ul>
  </div>
</div>
`;

exports[`GifGif submits a reason for the pronunciation 1`] = `
<div>
  <div class="rounded shadow text-grey-darker">
    <div class="py-8">
      <h3 class="text-center">How do you pronounce the word <span class="font-light uppercase">gif</span>?</h3>
      <div class="flex justify-center py-8">
        <label for="gif1" class="mx-8">
          GIF
          <input type="radio" id="gif1" value="gif1" name="pronunciation">
        </label>
        <label for="gif2" class="mx-8">
          GIF
          <input type="radio" id="gif2" value="gif2" name="pronunciation">
        </label>
      </div>
    </div>
    <div class="flex justify-center">
      <div class="w-3/4">
        <input id="reason" type="text" placeholder="Why do you think it should be pronounced that way?" class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight">
      </div>
    </div>
    <div class="flex justify-center my-4">
      <div class="w-3/4">
        <input id="name" type="text" placeholder="What is your name?" class="bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-purple rounded w-full py-2 px-4 text-grey-darker leading-tight">
      </div>
    </div>
    <div class="flex justify-center">
      <button id="submit" class="bg-blue hover:bg-blue-dark text-white font-bold py-2 px-4 rounded my-8 opacity-50 cursor-not-allowed" disabled="disabled">
        Submit
      </button>
    </div>
  </div>
  <div class="py-8">
    <h4 class="text-grey-darker text-center">User Responses:</h4>
    <ul>
      <li class="py-4">
        <p><span class="font-light uppercase">gif</span> — All the cool kids say it that way</p>
        <p class="text-sm mt-2">Submitted by <span class="uppercase">Samantha</span> on 2018-07-25</p>
      </li>
    </ul>
  </div>
</div>
`;
